File: farumasi_patient_portal/src/lib/consult-attachments.ts
/** Consult chat attachment helpers — keep product vs image vs file distinct. */

const PRODUCT_PATH = /\/(?:store|inventory|products)\/([^/?#]+)/i;
const UPLOAD_PATH = /\/uploads\//i;
const IMAGE_EXT = /\.(png|jpe?g|webp|gif|bmp|svg)(\?.*)?$/i;

export function isUploadAttachmentUrl(url: string): boolean {
  return UPLOAD_PATH.test(url);
}

export function isProductAttachmentUrl(url: string): boolean {
  if (isUploadAttachmentUrl(url)) return false;
  return PRODUCT_PATH.test(url);
}

/** Store route for a shared product, or undefined when not a product link. */
export function consultProductPath(url: string | null | undefined): string | undefined {
  if (!url || !isProductAttachmentUrl(url)) return undefined;
  const match = url.match(PRODUCT_PATH);
  if (!match) return undefined;
  return `/store/${match[1]}`;
}

export function productIdFromPath(url?: string | null): string | undefined {
  const path = consultProductPath(url ?? undefined);
  if (!path) return undefined;
  const match = path.match(/\/store\/([^/?#]+)/i);
  return match?.[1];
}

function isLikelyImageUrl(url: string): boolean {
  return (
    /^data:image\//i.test(url) ||
    /^blob:/i.test(url) ||
    IMAGE_EXT.test(url) ||
    /\/uploads\/(?:images?|media)\//i.test(url) ||
    (url.includes("res.cloudinary.com/") && /\/image\//i.test(url))
  );
}

export function inferAttachmentType(
  url: string | null | undefined,
  declared?: string | null,
): "image" | "file" | "product" | undefined {
  if (!url) return undefined;

  const normalizedDeclared =
    declared === "image" || declared === "file" || declared === "product"
      ? declared
      : undefined;

  if (isUploadAttachmentUrl(url)) {
    if (normalizedDeclared === "image" || normalizedDeclared === "file") {
      return normalizedDeclared;
    }
    return isLikelyImageUrl(url) ? "image" : "file";
  }

  if (isProductAttachmentUrl(url)) return "product";
  if (normalizedDeclared === "product") return "product";
  if (normalizedDeclared === "image") return isLikelyImageUrl(url) ? "image" : "file";
  if (normalizedDeclared === "file") return "file";
  if (isLikelyImageUrl(url)) return "image";
  return "file";
}

File: lib/utils/consult_attachments.dart
/// Consult chat attachment helpers — keep product vs image vs file distinct.

final _productPath = RegExp(r'/(?:store|inventory|products)/([^/?#]+)', caseSensitive: false);
final _uploadPath = RegExp(r'/uploads/', caseSensitive: false);
final _imageExt = RegExp(r'\.(png|jpe?g|webp|gif|bmp|svg)(\?.*)?$', caseSensitive: false);

bool isUploadAttachmentUrl(String? url) {
  if (url == null || url.isEmpty) return false;
  return _uploadPath.hasMatch(url);
}

bool isProductAttachmentUrl(String? url) {
  if (url == null || url.isEmpty || isUploadAttachmentUrl(url)) return false;
  return _productPath.hasMatch(url);
}

String? consultProductPath(String? url) {
  if (url == null || !isProductAttachmentUrl(url)) return null;
  final match = _productPath.firstMatch(url);
  if (match == null) return null;
  return '/store/${match.group(1)}';
}

String? productIdFromPath(String? url) {
  final path = consultProductPath(url);
  if (path == null) return null;
  final match = RegExp(r'/store/([^/?#]+)', caseSensitive: false).firstMatch(path);
  return match?.group(1);
}

bool _isLikelyImageUrl(String url) {
  return url.startsWith('data:image/') ||
      url.startsWith('blob:') ||
      _imageExt.hasMatch(url) ||
      RegExp(r'/uploads/(?:images?|media)/', caseSensitive: false).hasMatch(url) ||
      (url.contains('res.cloudinary.com/') && url.contains('/image/'));
}

String? inferAttachmentType(String? url, String? declared) {
  if (url == null || url.isEmpty) return declared;

  const allowed = {'image', 'file', 'product'};
  final normalizedDeclared = allowed.contains(declared) ? declared : null;

  if (isUploadAttachmentUrl(url)) {
    if (normalizedDeclared == 'image' || normalizedDeclared == 'file') {
      return normalizedDeclared;
    }
    return _isLikelyImageUrl(url) ? 'image' : 'file';
  }

  if (isProductAttachmentUrl(url)) return 'product';
  if (normalizedDeclared == 'product') return 'product';
  if (normalizedDeclared == 'image') {
    return _isLikelyImageUrl(url) ? 'image' : 'file';
  }
  if (normalizedDeclared == 'file') return 'file';
  if (_isLikelyImageUrl(url)) return 'image';
  return 'file';
}
